### PR TITLE
listener: Local rate limit listener filter

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message: added local rate limit listener filter to enable rate limit before TLS handshake and filter chain matching
Additional Description:
Risk Level: low
Testing: unit and integration test
Docs Changes: *local_rate_limiting.rst
Fixes #Issue: #25116

### CODE REVIEW

File: docs/root/intro/arch_overview/other_features/local_rate_limiting.rst
Comment (PeterL328) [inline docs/root/intro/arch_overview/other_features/local_rate_limiting.rst:8]: Wonder if we should write `and filter chain matching` to be more precise

Comment (JuniorHsu) [reply docs/root/intro/arch_overview/other_features/local_rate_limiting.rst:8]: filter chain is better. thanks!

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc
Comment (yanavlasov) [inline source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc:40]: I'm unsure I understand the need to explicitly reset the shared pointer in the destructor, as it will be reset during normal clean-up.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:38]: It's actually filter config. I saw the convention is on the same file with filter. Should I rename to `FilterConfig` to avoid confusion?

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc:40]: right we probably wont need to reset. I stole this from network local ratelimit but I thought reset is needed since the rate limiter is shared in network filter.

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc]: done. thanks!

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (yanavlasov) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:38]: Normally configuration and filter implementations are split into dedicated files. I would expect the config class to be in the config.h/cc files together with the factory.

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:38]: Factory initializes the filter, and filter needs the config information. 
That is, if we put factory and config into `config.h/cc`, we have circular dependency between config module and filter module, which fails the bazel check. 

I'm unsure the we have precedence putting config with factory. 
We have one case in http land `jwt_authn`, which put config into its own file.
However it needs to have another own file to handle factory to avoid circular dependency.

I could do three pairs of .h/.cpp if we really want to. 

```
 git grep "class FilterConfig" -- *.h
source/extensions/filters/http/alternate_protocols_cache/filter.h:17:class FilterConfig {
source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h:37:class FilterConfig {
source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h:67:class FilterConfigImpl : public FilterConfig {
source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h:61:class FilterConfig : public ::Envoy::Router::RouteSpecificFilterConfig {
source/extensions/filters/http/custom_response/config.h:30:class FilterConfig : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/ext_authz/ext_authz.h:54:class FilterConfig {
source/extensions/filters/http/ext_authz/ext_authz.h:242:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/ext_proc/ext_proc.h:68:class FilterConfig {
source/extensions/filters/http/ext_proc/ext_proc.h:108:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/grpc_http1_reverse_bridge/filter.h:69:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/jwt_authn/filter_config.h:41:class FilterConfig {
source/extensions/filters/http/jwt_authn/filter_config.h:62:class FilterConfigImpl : public Logger::Loggable<Logger::Id::jwt>,
source/extensions/filters/http/local_ratelimit/local_ratelimit.h:70:class FilterConfig : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/lua/lua_filter.h:375:class FilterConfig : Logger::Loggable<Logger::Id::lua> {
source/extensions/filters/http/lua/lua_filter.h:414:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/oauth2/filter.h:128:class FilterConfig {
source/extensions/filters/http/ratelimit/ratelimit.h:41:class FilterConfig {
source/extensions/filters/http/ratelimit/ratelimit.h:118:class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
source/extensions/filters/http/tap/tap_filter.h:37:class FilterConfig {
source/extensions/filters/http/tap/tap_filter.h:57:class FilterConfigImpl : public FilterConfig, public Extensions::Common::Tap::ExtensionConfigBase {
source/extensions/filters/http/wasm/wasm_filter.h:24:class FilterConfig : Logger::Loggable<Logger::Id::wasm> {
source/extensions/filters/network/wasm/wasm_filter.h:24:class FilterConfig : Logger::Loggable<Logger::Id::wasm> {
```

Comment (tyxia) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:38]: I think the question here is : Do you need this separate class`class config`? Also, looks like the `config` class doesn't use synchronization primitive or use TLS (i.e., RCU pattern). So, I am not sure thread safe is guaranteed here when this class  is not read-only. (canCreateConnection() is  fine as its underlying `requestAllowedHelper` use atomic operation)

I think it could just work if we remove config class and its members/methods are moved into to `class Filter`? The threading safe is also naturally achieved.  WDYT?

Comment (tyxia) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:38]: On second thought, it is good to have this `config` class as it is, since its members are read only and rate limiter is not shared.

Maybe rename it to 'FilterConfig' to be consistent with other places in code base.

**Review by tyxia — COMMENTED** [review]
Thanks for working on this! Left some comments around `config/FilterConfig` class

**Comment by phlax** [comment]
@JuniorHsu seems this needs a main merge and is awaiting some feedback

/wait

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h]: done. thanks

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h]: make sense. thanks! i'll fix network/local_ratelimit in a separated PR

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h]: done. thanks!

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (tyxia) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h]: I am not sure if we need to fix `network/local_ratelimit`. 

My understanding of `shared_ptr` there is 1) if `share_key` has value(i.e. sharing is enabled), they leverage shared_ptr 2) if `share_key` is empty(i.e., sharing is not enable), they just `make_shared` every time it is created, so each one has separate control with ref_count=1, which basically use it like a unique_ptr. 
And I think the reason `shared_ptr` is used there is because of coding convenience/consistence as you just need same type of object for above two cases.

Thank you for addressing my comments and adding tests! I will take a look later.

**Comment by yanavlasov** [comment]
LGTM from me. Waiting for final review from @tyxia

**Comment by JuniorHsu** [comment]
Hello @markdroth,
Given we're closed to finish the implementation review, could you have a review on the api part?
Generally speaking, this api is a simplified version of [network local rate limit](https://github.com/envoyproxy/envoy/blob/main/api%2Fenvoy%2Fextensions%2Ffilters%2Fnetwork%2Flocal_ratelimit%2Fv3%2Flocal_rate_limit.proto)

Thanks!

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (tyxia) [inline source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:69]: I am a bit confused, why we  override `onData` and `maxReadBytes` here as this filter inherits from `Network::ListenerFilter` which only has one `onAccept()`

File: test/extensions/filters/listener/local_ratelimit/local_ratelimit_test.cc
Comment (tyxia) [inline test/extensions/filters/listener/local_ratelimit/local_ratelimit_test.cc:45]: If you remove those unnecessary overrides from implementation, the function (like this one) and related tests  seems can also be removed?

File: test/extensions/filters/listener/local_ratelimit/local_ratelimit_integration_test.cc
Comment (tyxia) [inline test/extensions/filters/listener/local_ratelimit/local_ratelimit_integration_test.cc:69]: Just to confirm (for my own learning purpose): If you add a sleep(100000s) in this test, the token will be re-filled and then it will not be rate limited?

**Review by tyxia — COMMENTED** [review]
Just left a few comments(mostly nits), we are getting close to there. Thanks!

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.cc:8]: probably not i could give it a try to remove some but i don't have a good tool

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:55]: oh i misread that. bare object should be fine for listener filter

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:69]: since we don't have a pass through listener filter to inherit iirc. i could have one but that's should be a separated PR.

Comment (JuniorHsu) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:69]: i.e., all method in ListenerFilter are pure abstract and needs to be override

File: test/extensions/filters/listener/local_ratelimit/local_ratelimit_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/listener/local_ratelimit/local_ratelimit_test.cc:45]: same here.

File: test/extensions/filters/listener/local_ratelimit/local_ratelimit_integration_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/listener/local_ratelimit/local_ratelimit_integration_test.cc:69]: should be yeah. by the way we do enable in our prod and it refills which is great. and we already prove the filter can save handshake issue on scale (see the issue on description.

File: source/extensions/filters/listener/local_ratelimit/local_ratelimit.h
Comment (tyxia) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:69]: Sorry, I am trying github vscode codespace out.  Somehow, it only shows `virtual FilterStatus onAccept(ListenerFilterCallbacks& cb) PURE;` (same to all other inherited classes).

I checked the latests code. My comment should be ignored.

Comment (tyxia) [reply source/extensions/filters/listener/local_ratelimit/local_ratelimit.h:69]: ah, I create the codespace based off my fork repo that is [6590 commits behind](https://github.com/tyxia/envoy/compare/main...envoyproxy:envoy:main)... 
Sorry, please ignore this out-of-date comment

**Review by tyxia — APPROVED** [review]
LGTM. Thanks for good work!

**Comment by tyxia** [comment]
@mattklein123  Matt, as the code owner, maybe you can take a look and approve the API and PR if it looks good to you?

**Review by mattklein123 — APPROVED** [review]
API LGTM, thanks. LGTM to ship and iterate! Thank you.

**Comment by mattklein123** [comment]
Will defer to @yanavlasov for final/review merge.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
